libclamav_rust/src/alz.rs

```diff
@@ -0,0 +1,551 @@
+                }
+                2 => {
+                    self.compressed_size = u64::from(cursor.read_u16::<LittleEndian>()?);
+                    self.uncompressed_size = u64::from(cursor.read_u16::<LittleEndian>()?);
+                }
+                4 => {
+                    self.compressed_size = u64::from(cursor.read_u32::<LittleEndian>()?);
+                    self.uncompressed_size = u64::from(cursor.read_u32::<LittleEndian>()?);
+                }
+                8 => {
+                    self.compressed_size = cursor.read_u64::<LittleEndian>()?;
+                    self.uncompressed_size = cursor.read_u64::<LittleEndian>()?;
+                }
+                _ => return Err(Box::new(ALZParseError::new("Unsupported File Descriptor"))),
+            }
+        }
+
+        #[allow(clippy::cast_possible_truncation)]
+        let idx0: usize = cursor.position() as usize;
+        let idx1: usize = idx0 + self.head.file_name_length as usize;
+
+        if idx1 > cursor.get_ref().len() {
+            return Err(Box::new(ALZParseError::new("Invalid file name length")));
+        }
+
+        let filename = &cursor.get_ref().as_slice()[idx0..idx1];
+        cursor.set_position(idx1 as u64);
+
+        self.file_name = String::from_utf8_lossy(filename).into_owned();
+
+        if self.is_encrypted() {
+            cursor.read_exact(&mut self.enc_chk)?;
+        }
+
+        self.start_of_compressed_data = cursor.position();
+        cursor.set_position(self.start_of_compressed_data + self.compressed_size);
+
+        if self.start_of_compressed_data + self.compressed_size > cursor.get_ref().len() as u64 {
+            return Err(Box::new(ALZParseError::new(
+                "Invalid compressed data length",
+            )));
+        }
+
+        Ok(())
+    }
+
+    pub fn parse(&mut self, cursor: &mut std::io::Cursor<&Vec<u8>>) -> Result<(), ALZParseError> {
+        let result = self.parse_internal(cursor);
+        if result.is_err() {
+            let e = result.err().unwrap();
+            if e.is::<ALZParseError>() {
+                let ape: ALZParseError = *e.downcast::<ALZParseError>().unwrap();
+                return Err(ape);
+            }
+            return Err(ALZParseError::new("Not ALZ"));
+        }
+
+        Ok(())
+    }
+
+    pub fn is_supported(&self) -> Result<(), ALZUnsupportedError> {
+        if self.is_encrypted() {
+            return Err(ALZUnsupportedError::new(
+                "Encryption Unsupported".to_string(),
+            ));
+        }
+
+        if self.is_data_descriptor() {
+            return Err(ALZUnsupportedError::new(
+                "Data Descriptors are Unsupported".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /*
+     * This has no header/checksum validation.
+     */
+    fn extract_file_deflate(
+        &mut self,
+        cursor: &std::io::Cursor<&Vec<u8>>,
+        files: &mut Vec<ExtractedFile>,
+    ) -> Result<(), ALZExtractError> {
```

Comment:
```suggestion
    ) -> Result<(), Error> {
```

Reply:
Same deal here and later on with returning the error enum instead of specific type of error struct.